pkg/liqo-controller-manager/namespaceOffloading-controller/clusterselector_management.go

```diff
@@ -0,0 +1,85 @@
+package namespaceoffloadingctrl
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	k8shelper "k8s.io/component-helpers/scheduling/corev1"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	offv1alpha1 "github.com/liqotech/liqo/apis/offloading/v1alpha1"
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	liqoconst "github.com/liqotech/liqo/pkg/consts"
+)
+
+func (r *NamespaceOffloadingReconciler) selectCompliantVirtualNodes(noff *offv1alpha1.NamespaceOffloading,
+	clusterIDMap map[string]*mapsv1alpha1.NamespaceMap) error {
+	virtualNodes := &corev1.NodeList{}
+	if err := r.List(context.TODO(), virtualNodes,
+		client.MatchingLabels{liqoconst.TypeLabel: liqoconst.TypeNode}); err != nil {
+		klog.Error(err, " --> Unable to List all virtual nodes")
+		return err
+	}
+
+	// If here there are no virtual nodes is an error because it means that in the cluster there are NamespaceMap
+	// but not their associated virtual nodes
+	if len(virtualNodes.Items) == 0 {
+		err := fmt.Errorf(" No VirtualNodes at the moment in the cluster")
+		klog.Info(err)
+		return err
+	}
+
+	errorCondition := false
+	for i := range virtualNodes.Items {
+		match, err := k8shelper.MatchNodeSelectorTerms(&virtualNodes.Items[i], &noff.Spec.ClusterSelector)
+		if err != nil {
+			klog.Infof("%s -> Unable to offload the namespace '%s', there is an error in ClusterSelectorField",
+				err, noff.Namespace)
+			patch := noff.DeepCopy()
+			if noff.Annotations == nil {
+				noff.Annotations = map[string]string{}
+			}
+			noff.Annotations[liqoconst.SchedulingLiqoLabel] = fmt.Sprintf("Invalid Cluster Selector : %s", err)
```

Comment:
Shouldn't this error be reported as part of the status?

Reply:
Do you suggest having a new field in NamespaceOffloading resource? ATM we have only :
- name of the remote namespace
- global offloading status
- remote conditions for remote namesapces.

Reply:
At first sight, I would use the global offloading status (if the selector is invalid and assuming the resource to be immutable), than the offloading will certainly fail. Hence that could be set to Failed

Reply:
this controller does not have ownership of global offloading status; ATM  the Offloading Status controller (that reconcile NamespaceOffloading global status) manage this case of noClusterSelected as a global status of "NoClusterSelected", we can say this annotation is an extra information to let user understand the real cause of NotOffloading. Do you think this could be an acceptable approach ?

Reply:
Definitely, seems fine to me.